lib/lE_mirage.ml

```diff
@@ -0,0 +1,159 @@
+      Lwt_switch.turn_off stop_web_server >>= fun () -> Lwt.return v in
+    Lwt.both web_server provision_certificate >|= snd
+
+  let redirect config tls_port reqd =
+    let request = Httpaf.Reqd.request reqd in
+    let host =
+      match Httpaf.Headers.get request.Httpaf.Request.headers "host" with
+      | Some host -> host
+      | None -> Domain_name.to_string config.LE.hostname in
+    let response =
+      let port = if tls_port = 443 then None else Some tls_port in
+      let uri =
+        Fmt.str "https://%s%a%s" host
+          Fmt.(option ~none:nop (fmt ":%d"))
+          port request.Httpaf.Request.target in
+      let headers =
+        Httpaf.Headers.of_list [ ("location", uri); ("connection", "close") ]
+      in
+      Httpaf.Response.create ~headers `Moved_permanently in
+    Httpaf.Reqd.respond_with_string reqd response ""
+
+  let info =
+    let module R = (val Mimic.repr Paf.tls_protocol) in
+    let alpn_of_tls_connection (_edn, flow) =
+      match Paf.TLS.epoch flow with
+      | Ok { Tls.Core.alpn_protocol; _ } -> alpn_protocol
+      | Error _ -> None in
+    let peer_of_tls_connection (edn, _flow) = edn in
+    (* XXX(dinosaure): [TLS]/[ocaml-tls] should let us to project the underlying
+     * [flow] and apply [TCP.dst] on it.
+     * Actually, we did it with the [TLS] module. *)
+    let injection (_edn, flow) = R.T flow in
+    {
+      Alpn.alpn = alpn_of_tls_connection;
+      Alpn.peer = peer_of_tls_connection;
+      Alpn.injection;
+    }
+
+  let with_lets_encrypt_certificates ?(port = 443) stackv4v6 ~production config
+      handler =
+    let certificates = ref None in
+    let stop_http_server = Lwt_switch.create () in
+    let stop_alpn_server = Lwt_switch.create () in
+    let mutex = Lwt_mutex.create () in
+
+    let ctx =
+      let gethostbyname dns domain_name =
+        DNS.gethostbyname dns domain_name >>? fun ipv4 ->
+        Lwt.return_ok (Ipaddr.V4 ipv4) in
+      LE.ctx ~gethostbyname
+        ~authenticator:(Result.get_ok (Nss.authenticator ()))
+        (DNS.create stackv4v6) stackv4v6 in
+
+    let rec fill_certificates () =
+      LE.provision_certificate ~production config ctx >>= function
+      | Error _ as err ->
+          Lwt_switch.turn_off stop_http_server >>= fun () ->
+          Lwt_switch.turn_off stop_alpn_server >>= fun () -> Lwt.return err
+      | Ok v ->
+          Lwt_mutex.with_lock mutex (fun () ->
+              certificates := Some v ;
+              Lwt.return_unit)
+          >>= fun () ->
+          (* TODO(dinosaure): should we [reneg] all previous connections? *)
+          Time.sleep_ns (Duration.of_day 80) >>= fill_certificates in
+
+    let handshake tcp =
+      Lwt_mutex.with_lock mutex (fun () -> Lwt.return !certificates)
+      >>= function
+      | None -> Lwt.return_error `No_certificates
+      | Some certificates -> (
+          let cfg =
+            Tls.Config.server ~alpn_protocols:[ "http/1.1"; "h2" ] ~certificates
```

Comment:
I'm not sure if it's related to paf, but in my [example](https://github.com/kit-ty-kate/exn.st/blob/818ef15dbb3b3c8533927b7d7239c6cdf0fbc295/src/unikernel.ml#L30) HTTP/1.1 is always chosen over HTTP/2 for some reason:
```
$ curl -sI https://curl.se -o/dev/null -w '%{http_version}\n'
2
$ curl -sI https://exn.st -o/dev/null -w '%{http_version}\n'
1.1
```
The same behaviour can be observed in the web browser too.

Reply:
```suggestion
            Tls.Config.server ~alpn_protocols:[ "h2"; "http/1.1" ] ~certificates
```
I tried this way around instead and I can confirm that this was the bug. It seems that putting the http/1.1 version before causes the clients to totally ignore h2 (e.g. `curl --http2` always requests using http/1.1 instead of http/2, but the contrary with the new ordering is works just fine)

Reply:
Good catch!